File: problem-solving/baekjoon/1260_DFS와_BFS.py
# https://www.acmicpc.net/problem/1260
import sys

graph = []
visited = []

def dfs(v):
    visited[v] = True
    print("{} ".format(v), end="")

    for i in range(len(graph[v])):
        if graph[v][i] == 1 and not visited[i]:
            dfs(i)

def bfs(v):
    visited[v] = True
    queue = [v]

    while (len(queue) != 0):
        front = queue.pop(0)
        print("{} ".format(front), end="")
        for i in range(len(graph[v])):
            if graph[front][i] == 1 and not visited[i]:
                visited[i] = True
                queue.append(i)

def reset_visited():
    for i in range(len(visited)):
        visited[i] = False

if __name__ == "__main__":
    n, m, v = map(int, sys.stdin.readline().split())

    graph = [[0 for i in range(n+1)] for j in range(n+1)]
    visited = [False for i in range(n+1)]

    for i in range(m):
        vertex1, vertex2 = map(int, sys.stdin.readline().split())
        graph[vertex1][vertex2] = 1
        graph[vertex2][vertex1] = 1
    
    dfs(v)
    print()

    reset_visited()
    bfs(v)
    print()

File: problem-solving/baekjoon/1260_DFS와_BFS.go
// https://www.acmicpc.net/problem/1260
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	graph   [][]int
	visited []bool
	writer  *bufio.Writer
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, m, v int
	fmt.Fscanln(reader, &n, &m, &v)

	graph = make([][]int, n+1)
	for i := range graph {
		graph[i] = make([]int, n+1)
	}
	visited = make([]bool, n+1)

	for i := 0; i < m; i++ {
		var vertex1, vertex2 int
		fmt.Fscanln(reader, &vertex1, &vertex2)
		graph[vertex1][vertex2] = 1
		graph[vertex2][vertex1] = 1
	}

	dfs(v)
	fmt.Fprintln(writer, "")

	resetVisited()
	bfs(v)
	fmt.Fprintln(writer, "")
}

func dfs(v int) {
	visited[v] = true // dfs에 들어오면 방문했다고 판단
	fmt.Fprintf(writer, "%d ", v)

	for i := 0; i < len(graph[v]); i++ {
		if graph[v][i] == 1 && !visited[i] { // 연결되어 있고 방문하지 않은 경우
			dfs(i) // 재귀함수 호출
		}
	}
}

func bfs(v int) {
	visited[v] = true
	queue := []int{v}

	for len(queue) != 0 {
		front := queue[0]
		fmt.Fprintf(writer, "%d ", front)
		queue = queue[1:]
		for i := 0; i < len(graph[v]); i++ {
			if graph[front][i] == 1 && !visited[i] { // 연결되어 있고 방문하지 않은 경우
				visited[i] = true        // 방문 표시
				queue = append(queue, i) // 큐에 삽입
			}
		}
	}
}

func resetVisited() {
	for i := 0; i < len(visited); i++ {
		visited[i] = false
	}
}
